### DISCUSSION TITLE
Load font (awesome)

### DISCUSSION THREAD

**Discussion by elveejay1** [AvaloniaUI/Avalonia#8954] [Q&A] [answered]
In my previous WPF projects, I've had a static class to load fonts and characters, which I've used for [Font Awesome](https://fontawesome.com/start) 

I have

```

   public static class TextHelper
    {
        public static FontFamily FontAwesome5Regular { get; } = new FontFamily(new Uri("pack://application:,,,/My.MainApplication;component/Resources/"), "./#Font Awesome 6 Free Regular");

}

```

I would then bind to it with FontFamily="{x:Static local:TextHelper.FontAwesome5Solid}"

I hope I can still do the same in Avalonia? 

It dies with: UriFormatException: Invalid URI: Invalid port specified.

**Comment by Gillibald** [comment, marked as answer]
`new FontFamily("avares://MyAssembly/Resources#Font Awesome 6 Free Regular")`

**Reply by elveejay1** [reply]
Thanks. Hm, I came a bit further - now I get an exception about "'Could not create glyph typeface for: Font Awesome 6 Free Solid.'" The filename is called
Font Awesome 6 Free-Regular-400.otf

**Reply by elveejay1** [reply]
I also tried to move the fonts to a subfolder Fonts and then in my csproj add
```
    <AvaloniaResource Include="Resources\Fonts\**" />
```

which made it claim it could not find the Font Awesome otf file (even though it is there)

**Reply by elveejay1** [reply]
I think I'm doing the same as in FontFamilyTests.cs, although not quite it seems.

**Reply by elveejay1** [reply]
Found a solution - switch from OTF to TTF files. Then it worked, using the AvaloniaResource as mentioned and @Gillibald 's suggestion.

**Reply by elveejay1** [reply]
@Gillibald 
Another issue, if I may. While things are now running under Windows, I get "System.InvalidOperationException: could not create glyph typeface for: Font Awesome 6 Solid" when running it under Linux (Ubuntu 22.04 in a Hyper-V VM). 
Are things different in Linux???

Edit: I tried to use the otf, in case that worked better with linux. I then get a FileNotFoundException, though it seems to be looking for the right file at the right path.

**Reply by Gillibald** [reply]
What version of Avalonia are you using?

**Reply by elveejay1** [reply]
0.10.18

```

  <ItemGroup>
    <PackageReference Include="Avalonia" Version="0.10.18" />
    <PackageReference Include="Avalonia.Desktop" Version="0.10.18" />
    <!--Condition below is needed to remove Avalonia.Diagnostics package from build output in Release configuration.-->
    <PackageReference Condition="'$(Configuration)' == 'Debug'" Include="Avalonia.Diagnostics" Version="0.10.18" />
    <PackageReference Include="Avalonia.ReactiveUI" Version="0.10.18" />
    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
    <PackageReference Include="System.Configuration.ConfigurationManager" Version="6.0.1" />
    <PackageReference Include="XamlNameReferenceGenerator" Version="1.3.4" />
	<PackageReference Include="Avalonia.Xaml.Behaviors" Version="0.10.18" />
	<PackageReference Include="Avalonia.Xaml.Interactions" Version="0.10.18" />
	<PackageReference Include="Avalonia.Xaml.Interactivity" Version="0.10.18" />
	<PackageReference Include="Material.Icons.Avalonia" Version="1.1.10" />
	<PackageReference Include="MessageBox.Avalonia" Version="2.0.2" />

  </ItemGroup>

```

**Reply by elveejay1** [reply]
Just to make sure - @Gillibald , will you need anything else from me (for instance a zip with the solution etc)? Just in case there is anything else *I* can do to further the investigation.

**Reply by Gillibald** [reply]
Just in case the font name is handled differently on Linux you can specify the FontFamily like so: new FontFamily("avares://MyAssembly/Resources#Font Awesome")

Let me know if that works for you

**Reply by elveejay1** [reply]
I currently have them as

```
        public static FontFamily FontAwesome5Regular { get; } = new FontFamily("avares://MyAssembly/Resources/Fonts#Font Awesome 6 Free Regular");
        public static FontFamily FontAwesome5Solid { get; } = new FontFamily("avares://MyAssembly/Resources/Fonts#Font Awesome 6 Free Solid");

```
where the ttf files are placed in subdir Resources\Fonts\ relative to the MyAssembly.csproj

(Edit: and yes, it is silly the properties are called 5 when I use version 6, I know ;-) )

Edit2: I assume this is how it is supposed to be (and that I should not truncate the font name)?

**Reply by elveejay1** [reply]
I have tried both with ttf and otf versions. As the avares://... was what I already had, I can confirm that this does not work.

**Reply by elveejay1** [reply]
@Gillibald any news on this? I'd rather not be pushy, but having relied on this in my Windows version and seeing it _not_ working in Linux has brought me in a bit of a pickle (and not [Pickle Rick](https://en.wikipedia.org/wiki/Pickle_Rick)) ;-) 
If there is anything else I can do or try myself or possibly provide to ease the investigation, please let me know. I run in a Hyper-V VM with Ubuntu 22.04.
If more time for you/the Avalonia team is needed, I'll try to code around it for now (looking at other non-GUI tasks).

**Comment by sn4k3** [comment]
As another alternative see this library: https://github.com/Projektanker/Icons.Avalonia to place icons easely

**Reply by elveejay1** [reply]
While I prefer to add the font directly, this certainly also has merit, and offers an oportunity to add custom icons if needed I guess.

**Reply by sn4k3** [reply]
> offers an oportunity to add custom icons if needed I guess.

You can. But with fontawesome and material design icons i guess you don't need any more :D

**Reply by elveejay1** [reply]
Heh...almost. We might actually need a few specialized icons. But if I can, I hope to use it as a font if Jedi Master @Gillibald has a solution that works with Linux ;-)
